utils/appliance/__init__.py

```diff
@@ -249,37 +250,103 @@ def seal_for_templatizing(self):
+                    ip_addresses.add(ep.hostname)
+                else:
+                    ip_address = resolve_hostname(ep.hostname)
+                    if ip_address is not None:
+                        ip_addresses.add(ip_address)
+        return list(ip_addresses)
+
     @property
     def managed_providers(self):
-        """Returns a set of providers that are managed by this appliance
+        """Returns a set of provider crud objects of known providers managed by this appliance
 
-        Returns:
-            :py:class:`set` of :py:class:`str` - provider_key-s
+        Note:
+            Provider types have different ways of identification:
+              Infra = hostname / ip
+              Cloud = credentials
+              Containers = hostname / ip
+              Middleware = hostname / ip
+
+        Warning:
+            If name of the provider on the appliance differs from name in yamls, the name from the
+            appliance is used instead so that we can interact with the provider.
         """
-        ip_addresses = set([])
-
+        ems_table = self.db["ext_management_systems"]
         # Fetch all providers at once, return empty list otherwise
         try:
-            query_res = list(self._query_endpoints())
+            found_ems = list(self.db.session.query(ems_table))
         except Exception as ex:
             self.log.warning("Unable to query DB for managed providers: %s", str(ex))
             return []
 
-        for ipaddress, hostname in query_res:
-            if ipaddress is not None:
-                ip_addresses.add(ipaddress)
-            elif hostname is not None:
-                if re.match(r"^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$", hostname) is not None:
-                    ip_addresses.add(hostname)
-                else:
-                    ip_address = resolve_hostname(hostname)
-                    if ip_address is not None:
-                        ip_addresses.add(ip_address)
-        provider_keys = set([])
-        for provider_key, provider_data in conf.cfme_data.get("management_systems", {}).iteritems():
-            if provider_data.get("ipaddress", None) in ip_addresses:
-                provider_keys.add(provider_key)
-        return provider_keys
+        from utils.providers import list_providers
+        creds_table = self.db["authentications"]
+        prov_cruds = list_providers()
+        recognized_by_ip = ["InfraManager", "ContainerManager", "MiddlewareManager"]
+        recognized_by_creds = ["CloudManager"]
+        ignored_types = ["NetworkManager"]
+
+        found_cruds = set()
+        unrecognized_ems_names = set()
+        for ems in found_ems:
+            # Skip any provider types that we don't care about
+            if any(p_type in ems.type for p_type in ignored_types):
+                continue
+            found_crud = None
+            for prov in prov_cruds:
+                # Name check is authoritative
+                if ems.name == prov.name:
```

Comment:
This is for cases when provider is set up in the UI manually and automation is ran against the appliance... People have done that in the past, complaining that the provider is there but automation doesn't want to find it. This would work around the issue - we cannot say that we have found the provider but not found it at the same time, that's definitely not correct. We can either adjust on our side or we have to edit the name in CFME. One or the other.